File: 2415.reverse-odd-levels-of-binary-tree.cpp
#include <bits/stdc++.h>

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
    int lvl = 0;

public:
    // T: O(n)
    // S: O(n) recursion stack
    TreeNode* reverseOddLevels(TreeNode* root) {
        if (root == nullptr) {
            return root;
        }
        helper(root->left, root->right, 1);
        return root;
    }

private:
    void helper(TreeNode* l, TreeNode* r, int lvl) {
        if (l == nullptr || r == nullptr) {
            return;
        }
        if (lvl & 1) {
            std::swap(l->val, r->val);
        }
        ++lvl;
        helper(l->left, r->right, lvl);
        helper(l->right, r->left, lvl);
    }
};

File: 2415.reverse-odd-levels-of-binary-tree.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # BFS
    # T: O(n)
    # S: O(n)
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None
        
        q = deque()
        q.append(root)
        lvl = 0

        while len(q) > 0:
            if lvl & 1:
                l = 0
                r = len(q) - 1
                while l < r:
                    q[l].val, q[r].val = q[r].val, q[l].val
                    l += 1
                    r -= 1
            
            for _ in range(len(q)):
                cur = q.popleft()
                if cur.left is not None:
                    q.append(cur.left)
                if cur.right is not None:
                    q.append(cur.right)
            lvl += 1

        return root
